react-native/react/actions/login2.js

```diff
@@ -277,25 +260,144 @@ export function logout () {
+// Show a user/pass screen, call cb(user, passphrase) when done
+// title/subTitle to customize the screen
+export function askForUserPass (title, subTitle, cb, hidePass = false) {
+  return (dispatch) => {
+    dispatch(routeAppend({
+      parseRoute: {
+        componentAtTop: {
+          component: UserPass,
+          mapStateToProps: state => state.login2.userPass,
+          props: {
+            title,
+            subTitle,
+            hidePass,
+            onSubmit: (username, passphrase) => {
+              dispatch({
+                type: Constants.actionSetUserPass,
+                username,
+                passphrase
+              })
+
+              cb()
+            }
+          }
+        }
+      }
+    }))
+  }
+}
 
-      if (error) {
-        dispatch({
-          type: Constants.deviceNameSet,
-          error: true,
-          payload: error
-        })
-      } else {
+export function askForPaperKey (cb) {
+  return (dispatch) => {
+    dispatch(routeAppend({
+      parseRoute: {
+        componentAtTop: {
+          component: PaperKey,
+          mapStateToProps: state => state.login2,
+          props: {
+            onSubmit: (paperKey) => { cb(paperKey) }
+          }
+        }
+      }
+    }))
+  }
+}
+
+// Show a device naming page, call cb(deviceName) when done
+// existing devices are blacklisted
+export function askForDeviceName (existingDevices, cb) {
+  return (dispatch) => {
+    dispatch({
+      type: Constants.actionAskDeviceName,
+      existingDevices,
+      onSubmit: (deviceName) => {
         dispatch({
-          type: Constants.deviceNameSet
+          type: Constants.actionSetDeviceName,
+          deviceName
         })
 
-        // TODO multiple things do this, what do we do next? individual reducers?
+        cb()
       }
-    }, 1000)
+    })
+
+    dispatch(routeAppend({
+      parseRoute: {
+        componentAtTop: {
+          component: SetPublicName,
+          mapStateToProps: state => state.login2.deviceName
+        }
+      }
+    }))
+  }
+}
+
+export function registerWithUserPass () {
+  return (dispatch, getState) => {
+    const title = 'Registering with your Keybase passphrase'
+    const subTitle = 'lorem ipsum lorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsum'
+    const provisionMethod = enums.provisionUi.ProvisionMethod.passphrase
+    startLoginFlow(dispatch, getState, provisionMethod, title, subTitle, Constants.actionRegisteredWithUserPass)
   }
 }
+
+export function registerWithPaperKey () {
+  return (dispatch, getState) => {
+    const title = 'Registering with your paperkey requires your username'
+    const subTitle = 'Different lorem ipsum lorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsum'
+    const provisionMethod = enums.provisionUi.ProvisionMethod.paperKey
+    startLoginFlow(dispatch, getState, provisionMethod, title, subTitle, Constants.actionRegisteredWithPaperKey)
```

Comment:
i can ask before the whole thing happens but it'd be harder to get it to happen inbetween the rpcs. since i can front load it i'll make the change but i don't know how much it affects the user flow for the order to be slightly different